### DISCUSSION TITLE
Invalidation from Mutations

### DISCUSSION THREAD

**Discussion by reichhartd** [orval-labs/orval#506] [General]
Currently, we invalidate queries based on mutations according to the instructions on the [React Query website](https://tanstack.com/query/v4/docs/guides/invalidations-from-mutations).

We would like to keep it that way. To implement this with the automatic API generation I see two possibilities:

1. `MutationObserver`
   React Query gives you the possibility to observe mutations. As far as I understand, this is only properly possible if you provide a `mutationKey`. Orval does not use a `mutationKey` for mutations at the moment.

2. Override
   You use the `override` as follows:

    ```ts
                override: {
                    operations: {
                        UpdateUser: {
                            query: {
                                options: {
                                    onSuccess: async () => queryClient.invalidateQueries(getFindUsersQueryKey({})),
                                },
                            },
                        },
                    },
                },
    ```

However, there are several problems with this:

-   You can give `options` only for `queries`, no `mutations`.
-   Asynchronous functions are not generated correctly

    ```ts
        const query = useQuery<Awaited<ReturnType<typeof updateUser>>, TError, TData>(queryKey, queryFn, {
           enabled: !!userUuid,
           onSuccess: () =>.
               __async(void 0, null, function* () {
                   return queryClient.invalidateQueries(getFindUsersQueryKey({}));
               }),
           { queryOptions,
       }) as UseQueryResult<TData, TError> & { queryKey: QueryKey };
    ```

-   Certain imports like `queryClient` or `getFindUsersQueryKey` are missing. I can import them via `header`, but not for specific files. We use the `mode` `tags-split`, and therefore this `header` is used in every file, even if it is not necessary there. This generates TypeScript errors.

@anymaniax Did I miss a possibility? What would be your advice on which way I should go?

**Comment by anymaniax** [comment]
Hello @reichhartd, with the new orval@alpha you can now pass a `queryOptions` and `mutationOptions` which are mutators and that should help to solve your problem I think.

**Comment by daviesro** [comment]
@reichhartd Did you eventually solve this using the queryOptions or mutationOptions? And if so, do you happen to have a snippet to share? I’m in the same boat and have been looking for a way to invalidate queries on a POST

**Comment by reichhartd** [comment]
Hey @anymaniax @daviesro 
Unfortunately, this is not a priority for us at the moment, as we are working on another project. If that should change, I will try it out and post code snippets here.

**Comment by dhess** [comment]
@anymaniax Do you have an example of how the new `mutationOptions` is meant to be used in this way?
